Torchserve v2 protocol (#1870)
kserve/kserve: docs/samples/v1beta1/torchserve/README.md

## TorchServe with KFS envelope inference endpoints

> The KFServing/TorchServe integration supports KFServing v1 and v2 protocols.

Reviewer: ```suggestion
The KServe/TorchServe integration supports both v1 and v2 protocols.
```

V2 Protocol

| API  | Verb | Path | Payload |
| ------------- | ------------- | ------------- | ------------- |
| Predict  | POST  | /v2/models/<model_name>/infer  | Request: {"id" : $string #optional,"parameters" : $parameters #optional,"inputs" : [ $request_input, ... ],"outputs" : [ $request_output, ... ] #optional}  Response:{"model_name" : $string,"model_version" : $string #optional,"id" : $string,"parameters" : $parameters #optional,"outputs" : [ $response_output, ... ]}
| Explain  | POST  | /v2/models/<model_name>/explain  | Request: {"id" : $string #optional,"parameters" : $parameters #optional,"inputs" : [ $request_input, ... ],"outputs" : [ $request_output, ... ] #optional}  Response:{"model_name" : $string,"model_version" : $string #optional,"id" : $string,"parameters" : $parameters #optional,"outputs" : [ $response_output, ... ]} ||

[Sample requests for text and image classification](https://github.com/pytorch/serve/tree/master/kubernetes/kfserving/kf_request_json)

## Autoscaling
One of the main serverless inference features is to automatically scale the replicas of an `InferenceService` matching the incoming workload.

> KFServing by default enables [Knative Pod Autoscaler](https://knative.dev/docs/serving/autoscaling/) which watches traffic flow and scales up and down

Reviewer: ```suggestion
KServe by default enables [Knative Pod Autoscaler](https://knative.dev/docs/serving/autoscaling/) which watches traffic flow and scales up and down
```

based on the configured metrics.

[Autoscaling Example](autoscaling/README.md)

## Canary Rollout
Canary rollout is a deployment strategy when you release a new version of model to a small percent of the production traffic.

[Canary Deployment](canary/README.md)

## Monitoring
[Expose metrics and setup grafana dashboards](metrics/README.md)

## gRPC

#### Clone Torchserve Repo

```
git clone https://github.com/pytorch/serve
```

#### Install gRPC python dependencies

```
pip install -U grpcio protobuf grpcio-tools
```

#### Generate python gRPC client stub using the proto files

```bash
python -m grpc_tools.protoc --proto_path=serve/frontend/server/src/main/resources/proto/ --python_out=. --grpc_python_out=. serve/frontend/server/src/main/resources/proto/inference.proto serve/frontend/server/src/main/resources/proto/management.proto
```

#### Run gRPC Inference

```bash

> python torchserve_gprc_client.py infer <model_name> <input json>

Reviewer: ```suggestion
python torchserve_grpc_client.py infer <model_name> <input json>
```
